### PR TITLE
Added Discrete Scale to Plot Sub-Dialog

### LINKED ISSUE #1789: adding discrete scales options to plot sub dialog
By default ggplot ignores unused factor levels, but it should be easy to include them. e.g. for bar chart it can be done like this: http://stackoverflow.com/a/10835659/6424231
We should look into how we could have these options generally.

### PR DESCRIPTION
**Pull request by Vitalis95**
Replaces PR #8374 
Fixes #1789 
@rdstern , @N-thony , @lloyddewit , have a look at the progress made  so far on adding the discrete scale to plots sub-dialog,

### CODE REVIEW

**Review by rdstern — CHANGES_REQUESTED** [review]
@Vitalis95 well done.  This is already partly working ok.  The title part is fine.
The label part needs a different control.  I don't know whether we have this anywhere yet.  You can use it, as it stands, by giving strings, separated by commas.  But what we need is the Prepare > Recode > Recode Factor control:

image

Maybe check with Patrick, but I suggest having the frequencies does no harm and we could perhaps use this control.  Then it delivers the new names to R separated by commas.

It should not be used unless we ask for it.  So keep the checkbox and perhaps call it `New Labels`.

Then I wasn't sure how to use the Scales?  I will have a more detailed look and mayv=be you have ideas.

I liked the Position Top and Bottom and tried it.  It didn't remember for the next time.
I liked the Drop Unised levels and that seems to work fine.

**Comment by rdstern** [comment]
@Vitalis95 could you please discuss the next steps on this tab with @Patowhiz and @lilyclements.  It mostly already seems to be working, which is great, i.e.
image

This is our usual survey with village (4 levels) and I am just displaying 2 of them, and I have also changed their labels.  But this is not a practical way of giving the commands, when there are a lot of levels.

Exciting that this is really getting somewhere and showing how to manipulate factors within a graph!

Above I suggested that the new labels could use an adaptation of the Recode Factor control.  But the Limits could use an adaptation of the control in the Filter (sub) dialog.

image

So, I wonder if we could combine the two, so add a rename column here just before the ticks?  Here of course the default should be all the levels ticked?  

And just to add to your complications, here is the main (bar chart) dialog.

image

This gives the bar chart in Descending order and ideally that order would be reflected in the order of the factor levels in the sub-dialog.

I suggest adapting the filter control might be better, because that control loads much more efficiently than the others.  But just choose the simplest to adapt.  I guess that's a @Patowhiz discussion?

Now the breaks possibilities.  Here is a good stackoverflow site, called ["Display every nth value on a discrete scale." ](https://stackoverflow.com/questions/52919899/ggplot2-display-every-nth-value-on-discrete-axis)

This is something I have wanted for a long time.  For example in climatic I may have the year as a factor, but I just want to label every other year.  The solution uses a function and I hope we can too?  I guess that's a @lilyclements question.  We could still call it breaks and it is a checkbox with an up-down control, with default 1 and perhaps going up to 12.

There is also this neat trick for what it calls guide, but perhaps we call it Rows (the rows for the labels)

`scale_x_discrete(guide = guide_axis(n.dodge = 2)), `

That's also considered in the same stackoverflow.  Here is the result:

image


Finally - (so far), could there be a checkbox for this nice feature:

image

**Comment by Patowhiz** [comment]
@N-thony after looking into this issue with @Vitalis95, it's best we adopt the feature  by first having a custom control for editing the labels. This will simplify the task here.

**Comment by N-thony** [comment]
> @N-thony after looking into this issue with @Vitalis95, it's best we adopt the feature by first having a custom control for editing the labels. This will simplify the task here.

@Patowhiz that sounds good. We can have further discussion about it next week.

**Comment by rdstern** [comment]
@Vitalis95 and @N-thony I very much agree with the custom control.

However, that will take some time.  And I suggest it would be good to have a quick fix, possibly as follows:

image

In the figure above, the new custom control will take the place of the current labels and limits control.
In the stop-gap, could the Labels, be as now, perhaps wider.
And could the Limits move to the right, and becomer a similar checkbox, opening the Limits field.

They are usable, for now.  So could enable a version of this tab to be merged.

Then ask whether we could have the Breaks option on the left - which needs a function to run, see above.
And the Rows possibility as described above.  Each of these would be checkboxes I assume, so the radio buttons inside the Scales box are possibly no longer needed?  Perhaps the box is called Options instead?

**Comment by Vitalis95** [comment]
[https://stackoverflow.com/questions/35583844/keep-grid-lines-and-optional-tick-marks-when-using-breaks-ggplot](url)
The `labels` must be of the same length as` breaks`

**Comment by rdstern** [comment]
@Vitalis95 that's correct.  But I think we will usually just use 1 of them at a time.

When we use both now, then we will (for now) just give them as strings with a comma between each element.  Not worth checking on the size for now, because later it will come from the user control.

**Comment by Vitalis95** [comment]
> @Vitalis95 that's correct. But I think we will usually just use 1 of them at a time.
> 
> When we use both now, then we will (for now) just give them as strings with a comma between each element. Not worth checking on the size for now, because later it will come from the user control.

Great,  test it now

**Comment by rdstern** [comment]
@Vitalis95 I need to play with this now, before making the next suggestions.  I suggest this is a really important addition to R-Instat, both in its own right and because of the message it gives of R-Instat and also R.  Coping with factors in graphs is difficult, as is shown in the documentation.  So, making it easy is also difficult, but could be very worthwhile.

So no work for you yet, though I am keen to know how we might make progress on the new custom control?  But I refer here to 2 documents.  First is a "solution to one of the problems that there is sometimes.  That is long labels for the factors - what are the [options](https://www.andrewheiss.com/blog/2022/06/23/long-labels-ggplot/)?  

This is his best option:
image

Second is the standard reference for ggplot on the [position scales for discrete data](https://ggplot2.tidyverse.org/reference/scale_discrete.html).  

I really like their example, partly because it is easy in R-Instat, but I think it is the first one that is monly easy in our general dialog, and not in the special cases.  I am thinking that this case itself may be a good reason to move that dialog and give it more importance?

image

And here is the graph, in R-Instat - just as in the tidyverse documentation:

image


So, both the x-axis and y-axis are discrete.  How can we deal with the labels in lots of different ways.  I need to explore what options we could have on the left-hand side, to complement the new control.  Meantime I can consider what is possible with the new control, using the way you have made it.  Back with more suggestions early next week.

**Comment by rdstern** [comment]
I think I now have the scales I would like for this tab.
So, in the Scales box, at thew top of it add in 3 more checkboxes, in a single row.  default is unchecked.  They are labelled  `From`:, `To:` and `By`:  If checked they show an up-down control from 1 upwards, in steps of 1.  Initially no top limit.

image


a) If a factor is there (on the maoin dialog, then the top limit could become the number of levels of the factor.  And the default then could be 1 for From, and the number of levels for To and 1 for the By.

I have no idea yet of the code behind.  But here is an example:

image

The x-factor has 79 levels (from 1935 to 2013, and the new control coming on the right of the dialog will let you do similar things more visually.  (but more tediously if there are loads and loads of levels.)  Here, From = 27 To: = 56 By = 2, would give the graph from the 27th level so 1961 to  1990 with labels every second year.  
I am not sure of the code yet.  I'll have a look.

b) Then above Expand, add anoither checkbox with Label Wrap Labels:  Default unchecked.  If checked it opens an up-down control with default 10 (as in the example above) and in steps of 1, Lowere limit maybe 1 for now, but see what happens, and no upper limit.  We show an example above of this really nice feature.

**Comment by N-thony** [comment]
@Vitalis95 any progress?

**Comment by rdstern** [comment]
@Vitalis95 I would really like to have an improved version of this dialog in the next release - which is Wednesday.  Any chance of the improvements mentioned above or should we try to merge the current version?

**Comment by Vitalis95** [comment]
@rdstern , I am still in discussion with @N-thony  about this.

**Review by rdstern — APPROVED** [review]
@Vitalis95 thst seems promising, so far.  I tried with the survey data and the numeric use of factors seems to work.
When I tried with the diamonds data - getting a bar chart for carat - which has many levels - then I couldn't get it to work - it just didn't seem to include the code.  

I look forward to further work on this tab.  It is important.

**Comment by Vitalis95** [comment]
@rdstern . @lloyddewit  I will make more changes on separate pull request
If this can be merged first to avoid merge conflicts since it involves multiple dialogues, and other developers are also working on it

File: instat/dlgBarAndPieChart.vb [hunk 1]
```diff
@@ -408,6 +415,10 @@ Public Class dlgBarAndPieChart
         clsScaleSizeAreaFunction = New RFunction
         clsDummyFunction = New RFunction
         clsPointsFunction = New RFunction
+        clsConcantenateFunction = New RFunction
+        clsOperator1 = New ROperator
+        clsOperator2 = New ROperator
+        clsOperator = New ROperator
```
Comment (lloyddewit) [inline instat/dlgBarAndPieChart.vb:421]: more meaningful names?
```suggestion
        clsLeftBracketOperator= New ROperator
        clsRightBracketOperator= New ROperator
        clsSequenceOperator = New ROperator
```

File: instat/dlgBarAndPieChart.vb [hunk 2]
```diff
@@ -934,6 +975,16 @@ Public Class dlgBarAndPieChart
         ucrChkBacktoback.Enabled = If(Not ucrReceiverByFactor.IsEmpty AndAlso (rdoFrequency.Checked OrElse rdoValue.Checked), True, False)
     End Sub
 
+    Private Sub AddLevels()
+        If ucrVariablesAsFactorForBarChart.bSingleVariable AndAlso ucrVariablesAsFactorForBarChart.ucrSingleVariable.strCurrDataType = "factor" Then
+            clsLevelsFunction.AddParameter("x", ucrVariablesAsFactorForBarChart.GetVariableNames(False), bIncludeArgumentName:=False)
+            clsXScaleDiscreteFunction.AddParameter("limits", clsRFunctionParameter:=clsConcantenateFunction)
+
```
Comment (lloyddewit) [inline instat/dlgBarAndPieChart.vb:982]:
```suggestion
```

File: instat/ucrAxes.vb [hunk 3]
```diff
@@ -254,7 +320,7 @@ Public Class ucrAxes
         bControlsInitialised = True
     End Sub
 
-    Public Sub SetRCodeForControl(bIsXAxis As Boolean, Optional strNewAxisType As String = "continuous", Optional clsNewXYScaleContinuousFunction As RFunction = Nothing, Optional clsNewXYlabTitleFunction As RFunction = Nothing, Optional clsNewXYScaleDateFunction As RFunction = Nothing, Optional clsNewBaseOperator As ROperator = Nothing, Optional bReset As Boolean = False, Optional bCloneIfNeeded As Boolean = False)
+    Public Sub SetRCodeForControl(bIsXAxis As Boolean, Optional strNewAxisType As String = "continuous", Optional clsNewXYScaleContinuousFunction As RFunction = Nothing, Optional clsNewXYScaleDiscreteFunction As RFunction = Nothing, Optional clsNewXYlabTitleFunction As RFunction = Nothing, Optional clsNewXYScaleDateFunction As RFunction = Nothing, Optional clsNewBaseOperator As ROperator = Nothing, Optional clsNewOperator1 As ROperator = Nothing, Optional clsNewOperator2 As ROperator = Nothing, Optional bReset As Boolean = False, Optional bCloneIfNeeded As Boolean = False)
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:323]: What's the purpose of `clsNewOperator1` and `clsNewOperator2`?
Should we give them more meaningful names?

**Review by rdstern — DISMISSED** [review]
@Vitalis95 would like this to be merged.  @lloyddewit when the code is ok with you I am happy with that.  It is an important addition.  Though incomplete it is much better than nothing.

File: instat/dlgBarAndPieChart.vb [hunk 4]
```diff
@@ -585,6 +617,9 @@ Public Class dlgBarAndPieChart
         clsScaleColourViridisFunction.AddParameter("discrete", "TRUE", iPosition:=5)
         clsScaleFillViridisFunction.AddParameter("discrete", "TRUE", iPosition:=5)
 
+        'clsXScaleDiscreteFunction.AddParameter("limits", clsRFunctionParameter:=clsConcantenateFunction)
```
Comment (lloyddewit) [inline instat/dlgBarAndPieChart.vb:620]: delete? (or add 'todo' comment explaining why line left in)
```suggestion
```

File: instat/dlgBoxPlot.vb [hunk 5]
```diff
@@ -56,6 +58,9 @@ Public Class dlgBoxplot
     Private clsXScaleDateFunction As New RFunction
     Private clsYScaleDateFunction As New RFunction
 
+    Private clsOperator1 As New ROperator
+    Private clsOperator2 As New ROperator
+
```
Comment (lloyddewit) [inline instat/dlgBoxPlot.vb:63]: If we set `sdgPlots.SetRCode()` to have default parameters, then do we still need these variables?
```suggestion
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@Vitalis95 Thank you for this PR. 
If we set `sdgPlots.SetRCode()` to have default parameters, then it may not be necessary to change so many files.
Please could you check this? I'll then continue with my review (hopefully with fewer changed files).
thanks

**Comment by Vitalis95** [comment]
@rdstern , it works now with diamonds data(ordered factor). @lloyddewit have a look at it

File: instat/dlgBarAndPieChart.vb [hunk 6]
```diff
@@ -408,6 +418,13 @@ Public Class dlgBarAndPieChart
         clsScaleSizeAreaFunction = New RFunction
         clsDummyFunction = New RFunction
         clsPointsFunction = New RFunction
+        clsConcantenateFunction = New RFunction
+        clsLeftBracketOperator = New ROperator
+        clsRightBracketOperator = New ROperator
+        clsConcantenateFunction = New RFunction
+        clsLeftBracketOperator = New ROperator
+        clsRightBracketOperator= New ROperator
```
Comment (lloyddewit) [inline instat/dlgBarAndPieChart.vb:426]: duplicate of above?

File: instat/dlgBoxPlot.vb [hunk 7]
```diff
@@ -33,6 +33,8 @@ Public Class dlgBoxplot
     Private clsYlabFunction As New RFunction
     Private clsXScaleContinuousFunction As New RFunction
     Private clsYScaleContinuousFunction As New RFunction
+    Private clsYScaleDiscreteFunction As New RFunction
+    Private clsXScaleDiscreteFunction As New RFunction
```
Comment (lloyddewit) [inline instat/dlgBoxPlot.vb:37]: In a comment below for `sdgPlots.SetRCode()`, I suggested to make the `clsNewYScaleDiscreteFunction` and `clsNewXScaleDiscreteFunction` parameters optional (with defaults of GgplotDefaults.clsYScaleDiscreteFunction.Clone() and GgplotDefaults.clsXScaleDiscreteFunction.Clone() respectively).

If we do that, then can we revert all the changes in this file (and many of the other files below)?

I'm happy to dicsuss if needed.
```suggestion
```

File: instat/dlgBoxPlot.vb [hunk 8]
```diff
@@ -270,6 +272,8 @@ Public Class dlgBoxplot
         clsLabsFunction = GgplotDefaults.clsDefaultLabs.Clone()
         clsXScaleContinuousFunction = GgplotDefaults.clsXScalecontinuousFunction.Clone()
         clsYScaleContinuousFunction = GgplotDefaults.clsYScalecontinuousFunction.Clone()
+        clsYScaleDiscreteFunction = GgplotDefaults.clsYScaleDiscreteFunction.Clone()
+        clsXScaleDiscreteFunction = GgplotDefaults.clsXScaleDiscreteFunction.Clone()
```
Comment (lloyddewit) [inline instat/dlgBoxPlot.vb:276]:
```suggestion
```

File: instat/dlgBoxPlot.vb [hunk 9]
```diff
@@ -412,6 +416,7 @@ Public Class dlgBoxplot
     Private Sub cmdOptions_Click(sender As Object, e As EventArgs) Handles cmdOptions.Click, toolStripMenuItemPlotOptions.Click
         sdgPlots.SetRCode(clsNewOperator:=ucrBase.clsRsyntax.clsBaseOperator, clsNewYScalecontinuousFunction:=clsYScaleContinuousFunction, clsNewXScalecontinuousFunction:=clsXScaleContinuousFunction,
                                 clsNewXLabsTitleFunction:=clsXlabsFunction, clsNewYLabTitleFunction:=clsYlabFunction, clsNewLabsFunction:=clsLabsFunction, clsNewFacetFunction:=clsRFacetFunction,
+                                clsNewYScaleDiscreteFunction:=clsYScaleDiscreteFunction, clsNewXScaleDiscreteFunction:=clsXScaleDiscreteFunction,
```
Comment (lloyddewit) [inline instat/dlgBoxPlot.vb:419]:
```suggestion
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@Vitalis95 Thank you for the changes. I don't think all my previous comments are resolved. I added some new comments to try and clarify. I'm happy to discuss if needed.

P.S. Please don't set my comments to 'resolved' otherwise I lose track of which ones I've already checked. I will set the comments to resolved when I review the changes. Thanks

File: instat/dlgBarAndPieChart.vb [hunk 10]
```diff
@@ -77,6 +79,11 @@ Public Class dlgBarAndPieChart
     Private clsScaleSizeAreaFunction As New RFunction
     Private clsDummyFunction As New RFunction
     Private clsPointsFunction As New RFunction
+    Private clsConcantenateFunction As New RFunction
```
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/dlgBarAndPieChart.vb:82]: @vitalis this still needs resolving, thanks

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@Vitalis95 thank you for the changes, this simplifies the code changes.
there are still many `dlg*.vb` files (13 or 14 I think) in this PR that should not be here. Please revert the changes in these files (I think mainly just blank lines) so that these files no longer appear in this PR. 
I will then complete my review, just focussing on the remaining files. 
Thanks
P.S. there is also an unresolved comment about `clsConcantenateFunction`

File: instat/ucrAxes.vb [hunk 11]
```diff
@@ -133,6 +138,38 @@ Public Class ucrAxes
         ucrInputMinorBreaksFrom.AddQuotesIfUnrecognised = False
         ucrInputMinorBreaksFrom.SetLinkedDisplayControl(lblMinorBreaksFrom)
 
+        ucrChkLabelsDiscrete.SetText("Labels")
+        ucrChkLabelsDiscrete.AddParameterPresentCondition(True, "labels")
+        ucrChkLabelsDiscrete.AddParameterPresentCondition(False, "labels", False)
+        ucrChkLabelsDiscrete.AddToLinkedControls(ucrInputMajorBreaksLabelsDiscrete, {True}, bNewLinkedAddRemoveParameter:=True, bNewLinkedHideIfParameterMissing:=True, bNewLinkedChangeToDefaultState:=True, objNewDefaultState:="")
+        ucrInputMajorBreaksLabelsDiscrete.SetValidationTypeAsList()
+
+
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:147]:
```suggestion
```

File: instat/ucrAxes.vb [hunk 12]
```diff
@@ -133,6 +138,38 @@ Public Class ucrAxes
...
+        ucrChkLimitsFrom.AddToLinkedControls(ucrNudFrom, {True}, bNewLinkedAddRemoveParameter:=True, bNewLinkedHideIfParameterMissing:=True, bNewLinkedChangeToDefaultState:=True, objNewDefaultState:=1)
+        ucrNudFrom.SetMinMax(1, Integer.MaxValue)
+
+        ucrChkLimitsTo.SetText("To")
+        ucrChkLimitsTo.AddParameterPresentCondition(True, "to")
+        ucrChkLimitsTo.AddParameterPresentCondition(False, "to", False)
+        ucrChkLimitsTo.AddToLinkedControls(ucrNudTo, {True}, bNewLinkedAddRemoveParameter:=True, bNewLinkedHideIfParameterMissing:=True, bNewLinkedChangeToDefaultState:=True, objNewDefaultState:=2)
+        ucrNudTo.SetMinMax(1, Integer.MaxValue)
+
+
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:172]:
```suggestion
```

File: instat/ucrAxes.vb [hunk 13]
```diff
@@ -155,6 +192,7 @@ Public Class ucrAxes
         ucrInputUpperLimit.SetValuesToIgnore({"NA"})
         ucrInputUpperLimit.SetLinkedDisplayControl(lblUpperLimit)
 
+
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:195]:
```suggestion
```

File: instat/ucrAxes.vb [hunk 14]
```diff
@@ -281,13 +347,21 @@ Public Class ucrAxes
             ucrInputPosition.SetItems(New Dictionary(Of String, String)(GgplotDefaults.dctYPosition))
             ucrInputPosition.SetDefaultState("Left")
         End If
-        ucrInputAxisType.SetName(strAxisType)
+
+        If bIsX Then
+            ucrInputPositionDiscrete.SetItems(New Dictionary(Of String, String)(GgplotDefaults.dctXPosition))
+            ucrInputPositionDiscrete.SetDefaultState("Bottom")
+        ElseIf Not bIsX Then
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:354]:
```suggestion
        Else
```

File: instat/ucrAxes.vb [hunk 15]
```diff
@@ -281,13 +347,21 @@ Public Class ucrAxes
...
+            ucrInputPositionDiscrete.SetItems(New Dictionary(Of String, String)(GgplotDefaults.dctYPosition))
+            ucrInputPositionDiscrete.SetDefaultState("Left")
+        End If
 
         clsXYScaleDateLimitFunction = New RFunction
         clsXYScaleDateLimitFunction.SetRCommand("c")
         clsXYScaleDateLimitFunction.AddParameter("from", clsRFunctionParameter:=ucrDtpLowerLimit.ValueAsRDate(), iPosition:=0)
         clsXYScaleDateLimitFunction.AddParameter("to", clsRFunctionParameter:=ucrDtpLowerLimit.ValueAsRDate(), iPosition:=1)
 
+
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:364]:
```suggestion
```

File: instat/ucrAxes.vb [hunk 16]
```diff
@@ -306,6 +383,22 @@ Public Class ucrAxes
             clsLimitsFunction.SetRCommand("c")
         End If
 
+        If clsXYScaleDiscreteFunction.ContainsParameter("limits") AndAlso clsXYScaleDiscreteFunction.GetParameter("limits").clsArgumentCodeStructure IsNot Nothing Then
+            clsLimitsFunctionDiscrete = clsXYScaleDiscreteFunction.GetParameter("limits").clsArgumentCodeStructure
+        Else
+            'TODO move to ggplot defaults
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:389]: can we implement this 'todo' in this PR?

File: instat/ucrAxes.vb [hunk 17]
```diff
@@ -306,6 +383,22 @@ Public Class ucrAxes
...
+        Else
+            'TODO move to ggplot defaults
+            clsLimitsFunctionDiscrete = New RFunction
+            clsLimitsFunctionDiscrete.SetRCommand("c")
+        End If
+
+        If clsXYScaleDiscreteFunction.ContainsParameter("breaks") AndAlso clsXYScaleDiscreteFunction.GetParameter("breaks").clsArgumentCodeStructure IsNot Nothing Then
+            clsLimitsFunctionDiscrete = clsXYScaleDiscreteFunction.GetParameter("breaks").clsArgumentCodeStructure
+        Else
+            'TODO move to ggplot defaults
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:397]: can we implement this 'todo' in this PR?

File: instat/ucrAxes.vb [hunk 18]
```diff
@@ -320,6 +413,22 @@ Public Class ucrAxes
             clsMajorBreaksSeqFunction.SetRCommand("seq")
         End If
 
+        If clsXYScaleDiscreteFunction.ContainsParameter("breaks") Then
+            clsTempBreaksParam = clsXYScaleDiscreteFunction.GetParameter("breaks")
+            If clsTempBreaksParam.clsArgumentCodeStructure IsNot Nothing Then
+                clsMajorBreaksSeqDiscreteFunction = clsTempBreaksParam.clsArgumentCodeStructure
+            Else
+                'TODO move to ggplot defaults
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:421]: can we implement this 'todo' in this PR?

File: instat/ucrAxes.vb [hunk 19]
```diff
@@ -320,6 +413,22 @@ Public Class ucrAxes
...
+                clsMajorBreaksSeqDiscreteFunction = clsTempBreaksParam.clsArgumentCodeStructure
+            Else
+                'TODO move to ggplot defaults
+                clsMajorBreaksSeqDiscreteFunction = New RFunction
+                clsMajorBreaksSeqDiscreteFunction.SetRCommand("seq")
+            End If
+        Else
+            clsMajorBreaksSeqDiscreteFunction = New RFunction
+            clsMajorBreaksSeqDiscreteFunction.SetRCommand("seq")
+            clsMajorBreaksSeqDiscreteFunction.SetRCommand("seq")
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:428]: duplicates line above?
```suggestion
```

File: instat/ucrAxes.vb [hunk 20]
```diff
@@ -320,6 +413,22 @@ Public Class ucrAxes
...
+        End If
+
+
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:431]:
```suggestion
```

File: instat/ucrAxes.vb [hunk 21]
```diff
@@ -389,9 +498,35 @@ Public Class ucrAxes
         ucrInputMinorBreaksInStepsOf.SetRCode(clsMinorBreaksSeqFunction, bReset, bCloneIfNeeded:=bCloneIfNeeded)
 
         ucrChkLabels.SetRCode(clsXYScaleContinuousFunction, bReset, bCloneIfNeeded:=bCloneIfNeeded)
+
+
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:502]:
```suggestion
```

File: instat/ucrAxes.vb [hunk 22]
```diff
@@ -417,26 +552,40 @@ Public Class ucrAxes
...
 
-    Private Sub ucrPnlAxisTitle_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrPnlAxisTitle.ControlValueChanged, ucrInputTitle.ControlValueChanged
+    Private Sub ucrPnlAxisTitle_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrPnlAxisTitle.ControlValueChanged, ucrInputAxisType.ControlValueChanged
         SetLabel()
     End Sub
 
     Private Sub AddRemoveContinuousXYScales()
         If clsXYScaleContinuousFunction.clsParameters.Count > 0 Then
             clsBaseOperator.AddParameter("scale" & "_" & strAxis & "_" & "continuous", clsRFunctionParameter:=clsXYScaleContinuousFunction)
+
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:562]:
```suggestion
```

File: instat/ucrAxes.vb [hunk 23]
```diff
@@ -547,4 +740,52 @@ Public Class ucrAxes
...
+        ExpandControl()
+        AddRemoveScaleFunctions()
+        AddRemoveDiscreteXYScales()
+    End Sub
+
+    Private Sub ucrChkExpandDiscrete_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrChkExpandDiscrete.ControlValueChanged
+        ExpandControl()
+        AddRemoveScaleFunctions()
+        AddRemoveDiscreteXYScales()
+    End Sub
```
Comment (lloyddewit) [inline instat/ucrAxes.vb:766]: combine these 2 functions?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@Vitalis95 Thank you for the changes, this PR is now much simpler.
I think we can simplify the code changes further, please see suggestions.
Thanks

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern There have been extensive changes since your last approval. In theory, the functionality should be unchanged but a good retest is probably wise, thanks

**Review by rdstern — COMMENTED** [review]
@Vitalis95 see this test example:

image

This uses the standard diamonds data from the library.  And I was keen to present just the first 20 levels - not all of them.

I notice from the code here, there is nothing wrong, but it just isn't including anything to limit the levels?  When I use the ordinary survey data for the villages it does that fine.  I wonder what is not happening?

```
diamonds <- data_book$get_data_frame(data_name="diamonds")
last_graph <- ggplot2::ggplot(diamonds, mapping=ggplot2::aes(x=forcats::fct_infreq(f=as.factor(carat)), fill="")) + ggplot2::geom_bar(stat="count") + theme_grey() + ggplot2::theme(plot.caption=ggplot2::element_text(), plot.title=ggplot2::element_text(size=20), plot.subtitle=ggplot2::element_text(), axis.text.x=ggplot2::element_text())
data_book$add_object(data_name="diamonds", object_name="last_graph", object_type_label="graph", object_format="image", object=check_graph(graph_object=last_graph))

```

**Comment by Vitalis95** [comment]
@rdstern , have a look at it

**Review by rdstern — CHANGES_REQUESTED** [review]
@Vitalis95 that's much better.  

a) However I noticed that the checkboxes are separate, but you need to set both of them for anything to work.  If you just set one of them then it seems to give the whole thing.  Maybe the best is to have a single checkbox that makes both visible?  Up to you.
b) I then tried clicking on the Plot Options > bar chart and found that it has both the x and y blank.  If you do that and then return, you get an error.  This (I think) may be nothing to do with your current work, so I am copying @N-thony on this message.  I think it applies to all the pull-downs on this dialog and is a regression, from 7.6, where it did not occur.  Maybe you could check this feature or it could be a new issue.

**Comment by Vitalis95** [comment]
@rdstern , part b) is fixed in @EstherNjeriLiberatta  PR

**Comment by rdstern** [comment]
@Vitalis95 thanks - good spot and a great relief.  Must get that merged next week!

**Review by rdstern — CHANGES_REQUESTED** [review]
@Vitalis95 this now seems to work well for my examples of bar charts.
But before accepting I tried some more graphs and that seems to make it collapse.  I used the survey data for simplicity.
a) I did a line plot of yield by fert, so both continuous axes.  Then it crashes when getting plot options, because I think it is trying to get discrete all the time.  Now it should all be continuous.
b) Try also a line plot of yield by village which works - not nice plot, but that's not the point.  Then can I do anything with the xaxis (discrete) and yaxis (continuous)?

I am also not sure what to expect with the y-axis in a bar chart?

**Comment by rdstern** [comment]
@Vitalis95 here is an example you could be including.  It needs the general grapics dialog.  It uses the diamonds data from the ggplot2 library and does geom_jitter with both y and x a factor. (It uses a filter to just have carat > 1 so there are less points.)
Here is the dialog, with the data and the graph:

image

Now, both x and y are factors.  So, when using the plot options it should go to discrete for both x and y!
I hope you agree that's pretty nice.
:

**Comment by rdstern** [comment]
@Vitalis95 there are now build errors.  Perhaps this is a promising sign that you are in the midst of good changes?

**Pull request closed without merging**